serving/purgatorio/select-and-train-machine-learning-models/tools-and-libraries/tools-and-libraries.md

```diff
@@ -1,5 +1,321 @@
+
+[Scikit-Learn](https://scikit-learn.org/stable/) is one of the most complete, mature and well-documented libraries for Machine Learning tasks. It comes out-of-the-box with powerful and advanced models that offer facility functions for the data science process.
+
+Also, they have a [wonderful map](https://scikit-learn.org/stable/tutorial/machine_learning_map/index.html) to guide you in finding the right ML model for the problem you want to solve.
+
+Scikit-Learn resources:
+
+- [Getting Started](https://scikit-learn.org/stable/getting_started.html)
+- [Tutorials](https://scikit-learn.org/stable/tutorial/index.html)
+- [Examples](https://scikit-learn.org/stable/auto_examples/index.html)
+
+Also, Scikit-Learn is the library chosen by the awesome book [Hands-On Machine Learning with Scikit-Learn, Keras, and Tensorflow](https://www.amazon.it/Hands-Machine-Learning-Scikit-learn-Tensorflow/dp/1492032646/ref=pd_sbs_14_img_0/258-4192675-5285822?_encoding=UTF8&pd_rd_i=1492032646&pd_rd_r=b7c5b3eb-2cd0-4a82-a022-b68cefb4854f&pd_rd_w=I7oG4&pd_rd_wg=OyB1L&pf_rd_p=5b7cecc6-3c99-46e5-b3e1-0370a7bd6b14&pf_rd_r=AYXZXWBE4DBSAA0M723G&psc=1&refRID=AYXZXWBE4DBSAA0M723G), which is highly recommended to have in your pocket as a learning companion.
+
+Additional Machine Learning libraries and tools:
+
+- [A curated list of awesome Machine Learning frameworks, libraries, and software](https://github.com/josephmisiti/awesome-machine-learning#python)
+- [A curated list of awesome open source libraries to deploy, monitor, version and scale your machine learning models](https://github.com/EthicalML/awesome-production-machine-learning)
+
+
+## Deep Learning
+
+When it comes to Deep Learning it is not easy to decide which framework to recommend to a beginner. 
+
+Out there, on the Internet, at the query "which Deep Learning framework to learn first?", you quickly understand that the two most complete frameworks are [Tensorflow](https://www.tensorflow.org/) (from Google) and [Pytorch](https://pytorch.org/) (from Facebook), both open-source and full of useful features to work with Deep Learning.
+
+After evaluating several factors as:
+
+- completeness and maturity of the framework
+- usability and learning curve 
+- the clearness in the official documentation 
+- number of official and unofficial examples, guides and tutorials
+- industry-side diffusion 
+
+Virgilio suggests you start with learning TensorFlow and try Pytorch later.
+
+Brief argument: Tensorflow, especially from 2.0 version and with the [Keras](https://www.tensorflow.org/guide/keras) Interface (native in TF from 2.0), is a better choice for a beginner. Pytorch has different pros but it's more suitable for research and high flexibility needs.
+TF 2.0 is the tools chosen in the up-to-date version of the mythical (and _"mandatory"_) book [Hands-On Machine Learning with Scikit-Learn, Keras, and Tensorflow](https://www.amazon.it/Hands-Machine-Learning-Scikit-learn-Tensorflow/dp/1492032646).
+
```

Comment:
I would recommend Keras, Fast.ai, Pytorch in the order, as it's a good thing to learn high-level abstractions and then go deeper, unlike how Tensorflow has it. Tensorflow has sent early learner in a spin by their inside-out approach. :( it has been my personal experience as well.

Reply:
I can share my experience here https://github.com/neomatrix369/awesome-ai-ml-dl/blob/master/study-notes/README.md - my study notes, its a mix of Tensorflow and other topics but you can see however their 15 weeks course was hard to complete due to the introduction of low-level nitty-gritty topics in the beginning or too early on. I noticed when I moved to other methods of learning and other libraries or frameworks, I was able to complete them quicker without feeling slowed down.

Reply:
yes I will suggest Keras first, then TF. maybe I can add Pytorch but just as side. To learn keras and pytorch together it's a mess imho